src/cudamatrix/cu-matrix.h

```diff
@@ -334,6 +334,17 @@ class CuMatrixBase {
   void DiffLogSoftmaxPerRow(const CuMatrixBase<Real> &out_value,
                             const CuMatrixBase<Real> &out_deriv);
 
+  /// Add spatial regularization term to *this (`out_deriv`).
+  /// The regularization term is calculated from `out_value`.
+  /// Each row of `out_value` is first viewed as a 16*n grid,
+  /// convoluted with the 3x3 kernel, scaled with `scale`
+  /// and then added to `out_deriv`.
```

Comment:
I just noticed that for the derivative of the sqsum, you need to apply the high-pass filtering twice (and then multiply by -1.0 `*` scale).